src/compiler/checker.ts

```diff
@@ -7503,6 +7563,42 @@ namespace ts {
             return getUnionTypeFromSortedList(typeSet, aliasSymbol, aliasTypeArguments);
         }
 
+        function getUnionTypePredicate(signatures: ReadonlyArray<Signature>): TypePredicate {
+            let first: TypePredicate | undefined;
+            const types: Type[] = [];
+            for (let i = 0; i < signatures.length; i++) {

```

Comment:
why not for-of here?

Reply:
If we did #19309 we could enable `prefer-for-of` to catch this kind of error.